packages/native/types/base.d.ts

```diff
@@ -101,28 +101,36 @@ export interface CreateStyledComponent<
 > {
   /**
    * @typeparam AdditionalProps  Additional props to add to your styled component
+   * @typeparam InferredAdditionalProps  Should not be explicitly specified
```

Comment:
Could u explain what's happening here?

Reply:
there's an interesting bug in the following scenario:
```tsx
const Label = styled.label``
const Input = styled.input`
  & + ${Label}: {
    margin-left: 3px;
  }
`
```
`Label` is a `StyledComponent` but it since it's a function component it also matches the `FunctionInterpolation<Props>` interface. TypeScript sees this and thinks that the props accepted by the styled input component must match the props of any `FunctionInterpolation<Props>` provided to it, so the props accepted by `Input` end up being the intersection of those accepted by `input` *and* `label` elements. 

What my fix here does is make sure that the the props of a `FunctionInterpolation<Props>` is constrained by the props of `AdditionalProps` via an intermediate type `InferredAdditionalProps`, but the inferred type of `AdditionalProps` will not be affected by the props of the function passed in.

Reply:
> so the props accepted by Input end up being the intersection of those accepted by input and label elements.

I've just tested this on `next` branch and this doesn't happen there. How your changes have affected the inference algorithm for this? 🤔

Reply:
Could this be solved by using a conditional type within `Interpolation` type that would match our components as "strings" and thus not allowing them to be matched as `FunctionInterpolation<Props>`? Alternatively, we could type our components slightly differently - they are not truly `React.FC` so typing them as such is a stretch. However, from what I remember this just made things easier for us to type and a similar technique is used within `@types/react` to allow ExoticComponents to be handled by TS.

I've also thought that this should work:
```diff
diff --git a/packages/serialize/types/index.d.ts b/packages/serialize/types/index.d.ts
index 5fc08511..017731c9 100644
--- a/packages/serialize/types/index.d.ts
+++ b/packages/serialize/types/index.d.ts
@@ -41,7 +41,7 @@ export interface CSSObject
     CSSOthersObject {}
 
 export interface ComponentSelector {
-  __emotion_styles: any
+  __emotion_styles: string
 }
 
 export type Keyframes = {
@@ -56,6 +56,7 @@ export interface ArrayInterpolation<Props>
 
 export interface FunctionInterpolation<Props> {
   (props: Props): Interpolation<Props>
+  __emotion_styles?: never
 }
 
 export type Interpolation<Props> =
```
but apparently it doesnt.